app/src/main/java/net/aiscope/gdd_app/ui/sample_completion/SampleCompletionViewModel.kt

```diff
@@ -0,0 +1,179 @@
+    var usesGiemsa: Boolean = true
+    var giemsaFP: Boolean = true
+    var usesPbs: Boolean = true
+    var reusesSlides: Boolean = false
+    var bloodQuality: String = context.getString(R.string.spinner_empty_option)
+
+    //Fields for the microscope tab
+    var microscopeDamaged: Boolean = false
+    var microscopeMagnification: Int = 1000
+
+
+    override fun toString(): String {
+        val builder = StringBuilder()
+        this.javaClass.declaredFields.forEach {
+            builder.append(
+                "* " + it.name + " : " + this.javaClass.getDeclaredField(it.name).get(this) + "\n"
+            )
+        }
+        return builder.toString()
+    }
+
+    fun initVM() {
+        //TODO: inject dispatchers
+        viewModelScope.launch(Dispatchers.IO) {
+
+            val sample = repository.current()
+            disease = sample.disease
+            captures = sample.captures.completedCaptures
+
+            //So what we gonna do here for the meta stuff?
+            val lastMeta = repository.lastSaved()?.metadata
+            Timber.i("LAST META %s", lastMeta)
+            smearTypeId = MetadataMapper.getSmearTypeId(lastMeta?.smearType!!)
+            speciesValue = MetadataMapper.getSpeciesValue(context, lastMeta?.species!!)
+            comments = lastMeta?.comments
+
+            val lastMicroscopeQuality = repository.lastSaved()?.microscopeQuality
+            Timber.i("Last micro: %s", lastMicroscopeQuality)
+            microscopeDamaged = lastMicroscopeQuality?.isDamaged ?: false
+            microscopeMagnification = lastMicroscopeQuality?.magnification ?: 1000
+
+            //Set the values for the prep same as the last one
+            val lastPreparation = repository.lastSaved()?.preparation
+            Timber.i("Last prep %s", lastPreparation)
+
+            waterType = getWaterTypeValue(lastPreparation?.waterType)
+            usesGiemsa = lastPreparation?.usesGiemsa ?: true
+            giemsaFP = lastPreparation?.giemsaFP ?: true
+            usesPbs = lastPreparation?.usesPbs ?: true
+            reusesSlides = lastPreparation?.reusesSlides ?: false
+            bloodQuality = getBloodQualityValue(lastPreparation?.bloodQuality)
+        }
+    }
+
+    fun save() {
+        //TODO: inject dispatchers
+        viewModelScope.launch(Dispatchers.IO) {
+            val newQualityValues = MicroscopeQuality(microscopeDamaged, microscopeMagnification)
+            val newPreparation = SamplePreparation(
+                getWaterType(waterType),
+                usesGiemsa,
+                giemsaFP,
+                usesPbs,
+                reusesSlides,
+                getBloodQuality(bloodQuality)
+            )
+
+            //Aight so what else for the meta stuff?
+            val newMeta = SampleMetadata(
+                //TODO: move these methods out of MDMapper
+                smearType = MetadataMapper.getSmearType(smearTypeId!!),

```

Comment:
same here and the line under this.

Reply:
Changed the mapper to take nullable types